ufl/algorithms/apply_derivatives.py

```diff
@@ -1802,38 +2162,61 @@ def jacobian(self, o):
         return self.independent_terminal(o)
 
 
-class CoordinateDerivativeRuleDispatcher(MultiFunction):
+class CoordinateDerivativeRuleDispatcher(DAGTraverser):
     """Dispatcher."""
 
-    def __init__(self):
+    def __init__(self, compress=True, visited_cache=None, result_cache=None) -> None:
         """Initialise."""
-        MultiFunction.__init__(self)
-        self.vcache = defaultdict(dict)
-        self.rcache = defaultdict(dict)
+        super().__init__(compress=compress, visited_cache=visited_cache, result_cache=result_cache)
+        self._dag_traverser_cache = {}
+
+    @singledispatchmethod
+    def process(self, o: Expr) -> Expr:
+        """Process ``o``.
 
-    def terminal(self, o):
+        Args:
+            o: `Expr` to be processed.
+
+        Returns:
+            Processed object.
+
+        """
+        raise AssertionError(f"UFL expression expected: got {o}")
+
+    @process.register(Expr)
+    @process.register(BaseForm)
+    def _(self, o: Union[Expr, BaseForm]) -> Union[Expr, BaseForm]:
+        """Apply to expr and base form."""
+        return self.reuse_if_untouched(o)
+
+    @process.register(Terminal)
+    def _(self, o: Expr) -> Expr:
         """Apply to a terminal."""
         return o
 
-    def derivative(self, o):
+    @process.register(Derivative)
+    def _(self, o: Expr) -> Expr:
         """Apply to a derivative."""
         raise NotImplementedError(f"Missing derivative handler for {type(o).__name__}.")
 
-    expr = MultiFunction.reuse_if_untouched
-
-    def grad(self, o):
+    @process.register(Grad)
+    def _(self, o: Expr) -> Expr:
         """Apply to a grad."""
         return o
 
-    def reference_grad(self, o):
+    @process.register(ReferenceGrad)
+    def _(self, o: Expr) -> Expr:
         """Apply to a reference_grad."""
         return o
 
-    def coefficient_derivative(self, o):
+    @process.register(CoefficientDerivative)
+    def _(self, o: Expr) -> Expr:
         """Apply to a coefficient_derivative."""
         return o
 
-    def coordinate_derivative(self, o, f, w, v, cd):
+    @process.register(CoordinateDerivative)
+    @DAGTraverser.postorder_only_children([0])
+    def _(self, o: Expr, f: Expr) -> Expr:
```

Comment:
I guess a test on coordinate-derivative should yield an `*args` that is non-empty.

Reply:
`CoordinateDerivative ` still only needs `args=()`, I think.

Reply:
Yeah, I only looked at the signatures of the process method, but for the traversal of the graph, it seems like we never use args.